lessons/4.mdx

```diff
@@ -0,0 +1,818 @@
+    await contract.deployed()
+  })
+
+  describe('constructor', async () => {
+    it('set proper collection name', async function () {
+      const name = await contract.name()
+      expect(name).to.equal('TierNFT')
+    })
+
+    it('set proper collection symbol', async function () {
+      const symbol = await contract.symbol()
+      expect(symbol).to.equal('Tier')
+    })
+  })
+})
+```
+
+At the top we see constants with the name and symbol of the NFT collection:
+
+```javascript
+const CONTRACT_NAME = 'TierNFT'
+const COLLECTION_NAME = 'TierNFT'
+const COLLECTION_SYMBOL = 'Tier'
+```
+
+Next is a `describe('TierNFT', function () { ...` line. What's cool about
+`describe()` is that it logically groups tests together so you'll see nicely
+organized output.
+
+When we run the tests we see:
+
+```bash
+TierNFT
+    constructor
+      ✔ set proper collection name
+      ✔ set proper collection symbol
+```
+
+Notice "TierNFT" comes from the first `describe()` block and `constructor` is
+from the inner `describe()` block. The tests are then indented under there.
+Makes for a nicely organized suite of tests.
+
+Next notice the `beforeEach()` code block. When tests run it's important that
+the contract is deployed from scratch for each test and this is where that
+happens. "Before each" test, this code will run and re-inialize everything,
+which prevents one test's results from interfering with another test.
+
+What's happening in `beforeEach()`?
+
+The hardhat framework give us the tools to deploy our contract to a
+locally-running blockchain. So when this code finishes, we have a fully deployed
+contract, and a couple of wallets to use for testing -- `owner` is who has all
+the permissions since it deployed the contract, and `otherUser` is a wallet with
+less permissions -- these will be handy when testing `withdraw` later on.
+
+Those `let` statements above are to hold these values so we can use them
+throughout all our tests.
+
+Finally we then come across our first two actual tests:
+
+```javascript
+it('set proper collection name', async function () {
+  const name = await contract.name()
+  expect(name).to.equal('TierNFT')
+})
+
+it('set proper collection symbol', async function () {
+  const symbol = await contract.symbol()
+  expect(symbol).to.equal('Tier')
+})
+```
+
+They each call a method and verify the results. "Is the contract name really
+TierNFT?" "Is the contract symbol really Tier?"
+
+<ContentCallout emoji="💡" size="md" variant="info">
+  You'll notice a lot of `await` statements. That's because most of these calls
+  are asynchronous. Without `await` the program would keep going without waiting
+  for the result. You may want to take a detour and learn more about what those
+  do, but for now, think of them as basically waiting for an asynchronous call
+  to return before moving on to the next line.
+</ContentCallout>
+
+They also lead us to talk about general testing concepts.
```

Comment:
this sentence is a bit out of place visually, escpecially cos it has the word **also** in it.
 Should it be along with the sentence before the callout?

Reply:
Yeah, awkward sentence. I think I can just remove it in this case.